### PR TITLE
Treat ResourceName as string on Client class surface

### PR DESCRIPTION
**Pull request by andreamlin**
Addresses https://github.com/googleapis/toolkit/issues/1804.

### CODE REVIEW

**Comment by andreamlin** [comment]
PTAL

File: src/main/java/com/google/api/codegen/viewmodel/StaticLangApiMessageView.java [hunk 1]
```diff
@@ -75,11 +74,15 @@ public boolean hasRequestBody() {
   public abstract boolean hasRequiredProperties();
 
   // List of ResourceNames that this message contains.
-  public abstract List<RequestObjectParamView> resourceNames();
+  @Nullable
+  public abstract RequestObjectParamView resourceName();
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/viewmodel/StaticLangApiMessageView.java]: I'm a little confused about this. I can see how a request message would have a primary resource name, but couldn't there be additional fields that also have resource names?

File: src/main/resources/com/google/api/codegen/java/resource_name.snip [hunk 2]
```diff
@@ -10,13 +10,15 @@
 @end
 
 @private nameClassDef(resourceName)
-  public final class {@resourceName.typeName} implements ResourceName {
+  public final class {@resourceName.typeName} implements ResourceName, Struct {
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/resource_name.snip]: What's the motivation for adding `Struct` here?

File: src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java [hunk 3]
```diff
@@ -180,6 +181,18 @@ public static Name getResponseName(Method method) {
     return null;
   }
 
+  @Nullable
+  public DiscoveryField getResponseField(Method method) {
+    if (method.response() != null) {
+      Schema responseSchema =
+          method.response().reference() != null
+              ? method.response()
+              : method.getDocument().schemas().get(method.response().getIdentifier());
```
[the remark this answers is not in the document]
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java]: Sure.

File: src/main/java/com/google/api/codegen/viewmodel/StaticLangApiMessageView.java (hunk 1 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/viewmodel/StaticLangApiMessageView.java]: I haven't yet seen multiple resource names for a given Message object in a discovery document, so I don't know what that would look like.

**Comment by andreamlin** [comment]
PTAL

File: src/main/java/com/google/api/codegen/GeneratorVersionProvider.java [hunk 4]
```diff
@@ -24,14 +23,15 @@
...
+    return "0.0.5";
+    //    try {
+    //      properties.load(
+    //          GeneratorVersionProvider.class
+    //              .getResourceAsStream("/com/google/api/codegen/codegen.properties"));
+    //      version = properties.getProperty("version");
+    //    } catch (IOException e) {
+    //      e.printStackTrace(System.err);
+    //    }
+    //    return version;
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/GeneratorVersionProvider.java]: We should just delete this code. The right way to get the version is through a metadata file.

File: src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java [hunk 5]
```diff
@@ -167,15 +168,17 @@ public static Name getInterfaceName(String defaultInterfaceName) {
     return Name.anyCamel(Inflector.singularize(resource));
   }
 
-  /** Get the response type name from a method if the method has a non-null response type. */
+  /** Get the response type from a method if the method has a non-null response type. */
   @Nullable
-  public static Name getResponseName(Method method) {
+  public DiscoveryField getResponseField(Method method) {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java]: I'm a little confused - why would a response be a field? Doesn't a field have to be part of a message?

File: src/main/java/com/google/api/codegen/discogapic/transformer/java/JavaDiscoGapicRequestToViewTransformer.java [hunk 6]
```diff
@@ -245,21 +241,51 @@ private StaticLangApiMessageView generateRequestClass(
...
+    String resourceName = DiscoGapicNamer.getResourceIdentifier(discoMethod.path()).toLowerCamel();
+    StringBuilder description =
+        new StringBuilder(discoMethod.parameters().get(resourceName).description());
+    description.append(String.format("\nIt must have the format `%s`. ", discoMethod.path()));
+    description.append(String.format("\\`{%s}\\` must start with a letter,\n", resourceName));
+    description.append(
+        "and contain only letters (\\`[A-Za-z]\\`), numbers (\\`[0-9]\\`), dashes (\\`-\\`),\n"
+            + "     * underscores (\\`_\\`), periods (\\`.\\`), tildes (\\`~\\`), plus (\\`+\\`) or percent\n"
+            + "     * signs (\\`%\\`). It must be between 3 and 255 characters in length, and it\n"
+            + "     * must not start with \\`\"goog\"\\`.");
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/discogapic/transformer/java/JavaDiscoGapicRequestToViewTransformer.java]: I don't see this showing up in the baseline, did it not get piped through correctly?

File: src/main/java/com/google/api/codegen/transformer/StaticLangResourceObjectTransformer.java [hunk 7]
```diff
@@ -108,12 +108,14 @@ public RequestObjectParamView generateRequestObjectParam(
     if (!isRequired) {
       param.optionalDefault(namer.getOptionalFieldDefaultValue(fieldConfig, context));
     }
-    List<FieldCopyView> fieldCopyViews = new ArrayList<>();
+    List<StaticMemberView> fieldCopyViews = new ArrayList<>();
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/StaticLangResourceObjectTransformer.java]: The name implies that this refers to "static members," which doesn't quite seem right.

Also the variable name doesn't match the type now.

File: src/main/java/com/google/api/codegen/discogapic/transformer/java/JavaDiscoGapicRequestToViewTransformer.java (hunk 6 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/discogapic/transformer/java/JavaDiscoGapicRequestToViewTransformer.java]: The description comments aren't yet rendered by the snippets.

File: src/main/java/com/google/api/codegen/GeneratorVersionProvider.java (hunk 4 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/GeneratorVersionProvider.java]: Reverting this back to master. It was just a hack for when intellij couldn't find the properties file.

File: src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java (hunk 5 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java]: The response types in GCE are message types.

File: src/main/java/com/google/api/codegen/transformer/StaticLangResourceObjectTransformer.java (hunk 7 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/transformer/StaticLangResourceObjectTransformer.java]: I renamed the StaticMemberView class to StaticLangMemberView class, and also renamed that local var.

File: src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java (hunk 5 above)
Comment (garrettjonesgoogle) [reply src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java]: It's the terminology that's throwing me off - a method's attributes (e.g. request type, response type) aren't generally called "fields."

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline [hunk 8]
```diff
@@ -74,6 +91,20 @@ public final class AddressName implements ResourceName {
   }
 
 
+  @Override
```
Comment (andreamlin) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: Remove this method.

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline [hunk 9]
```diff
@@ -74,6 +91,20 @@ public final class AddressName implements ResourceName {
...
+  public Map<String, String> getFieldValues() {
+    Map<String, String> fieldMap = new HashMap<>();
+    fieldMap.put("address", address);
+    fieldMap.put("project", project);
+    fieldMap.put("region", region);
+    return fieldMap;
+  }
+
+  @Override
+  public AddressName parseFrom(String formattedString) {
```
Comment (andreamlin) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: Remove this method.

File: src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java (hunk 5 above)
Comment (andreamlin) [reply src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java]: Ok, renamed to getResponseType().

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline [hunk 10]
```diff
@@ -506,9 +592,44 @@ public final class Address implements ApiMessage {
...
+    if (fieldName.equals("selfLink")) {
+      return String.valueOf(selfLink);
+    }
+    if (fieldName.equals("status")) {
+      return String.valueOf(status);
+    }
+    if (fieldName.equals("users")) {
+      return String.valueOf(users);
+    }
+    throw new IOException(String.format("Field '%s' not found in Address fields", fieldName));
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: I think we should follow the semantics of maps in Java and return null.

Comment (andreamlin) [reply src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: Sure thing. There just wouldn't be a way to differentiate between an actual null value for that fieldName vs the fieldName being an incorrect field name.

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline [hunk 11]
```diff
@@ -407,8 +482,8 @@ package com.google.cloud.simplecompute.v1;
 
 import com.google.api.core.BetaApi;
 import com.google.api.gax.httpjson.ApiMessage;
-import com.google.common.collect.ImmutableList;
-import java.io.Serializable;
+import com.google.common.collect.ImmutableMap;
+import java.io.IOException;
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: Remove the IOException import now

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline [hunk 12]
```diff
@@ -905,9 +1017,29 @@ public final class AddressList implements ApiMessage {
     return fieldMap;
   }
 
+  @Override
+  public String getFieldStringValue(String fieldName) {
+    if (fieldName.equals("id")) {
+      return String.valueOf(id);
+    }
+    if (fieldName.equals("items")) {
+      return String.valueOf(items);
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: I don't think `String.valueOf` will return anything useful for lists... (you can solve this problem later, just add a TODO)

File: src/main/resources/com/google/api/codegen/java/resource_name_factory.snip [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+  }
+
+@end
+
+@private interfaceMethods(factory)
+  @@Override
+  public {@factory.resourceNameTypeName} parseFrom(String formattedString) {
+    return {@factory.resourceNameTypeName}.parse(formattedString);
+  }
+@end
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/resource_name_factory.snip]: add a newline

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline (hunk 12 above)
Comment (andreamlin) [reply src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: Done.

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline (hunk 11 above)
Comment (andreamlin) [reply src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: Done.

File: src/main/resources/com/google/api/codegen/java/resource_name_factory.snip (hunk 13 above)
Comment (andreamlin) [reply src/main/resources/com/google/api/codegen/java/resource_name_factory.snip]: Done.

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline [hunk 14]
```diff
@@ -372,18 +450,15 @@ public final class RegionName implements ResourceName {
...
 @Generated("by GAPIC")
 @BetaApi
-public final class RegionNameType implements ResourceNameType {
-  private static RegionNameType instance = new RegionNameType();
-
-  private RegionNameType() {}
-
-  public static RegionNameType instance() {
-    return instance;
+public final class RegionNameFactory implements ResourceNameFactory {
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: We don't need this given that a factory for `RegionName` can be acquired from `RegionName.newFactory()`, right?

File: src/main/java/com/google/api/codegen/discogapic/transformer/java/JavaDiscoGapicRequestToViewTransformer.java [hunk 15]
```diff
@@ -299,17 +325,17 @@ private StaticLangApiMessageView schemaToParamView(
...
     return paramView.build();
   }
 
   private void addApiImports(ImportTypeTable typeTable) {
     typeTable.getAndSaveNicknameFor("com.google.api.core.BetaApi");
-    typeTable.getAndSaveNicknameFor("com.google.common.collect.ImmutableList");
+    typeTable.getAndSaveNicknameFor("com.google.common.collect.ImmutableMap");
     typeTable.getAndSaveNicknameFor("com.google.api.gax.httpjson.ApiMessage");
-    typeTable.getAndSaveNicknameFor("java.io.Serializable");
+    typeTable.getAndSaveNicknameFor("java.io.IOException");
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/discogapic/transformer/java/JavaDiscoGapicRequestToViewTransformer.java]: Remove this

Comment (andreamlin) [reply src/main/java/com/google/api/codegen/discogapic/transformer/java/JavaDiscoGapicRequestToViewTransformer.java]: Done.

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline (hunk 14 above)
Comment (andreamlin) [reply src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: Right! Done.

**Comment by andreamlin** [comment]
Done when https://github.com/googleapis/gax-java/pull/481 gets merged in.

**Pull request merged**